Hudl.Mjolnir/Config/BulkheadConfigurationObserver.cs

```diff
@@ -0,0 +1,48 @@
+﻿using System;
+
+namespace Hudl.Mjolnir.Config
+{
+    internal sealed class BulkheadConfigurationObserver<T> : IObserver<BulkheadConfiguration>
+    {
+        private T _currentValue;
+        private readonly Func<BulkheadConfiguration, T> _expression;
+        private readonly Action<T> _onChange;
+        internal BulkheadConfigurationObserver(BulkheadConfiguration currentConfig, Func<BulkheadConfiguration, T> propertyToCheck, Action<T> onChange)
+        {
+            _expression = propertyToCheck;
+            _currentValue = _expression(currentConfig);
+            _onChange = onChange;
+        }
+
+        public void OnCompleted()
+        {
+            throw new NotImplementedException();
+        }
+
+        public void OnError(Exception error)
+        {
+            throw new NotImplementedException();
+        }
+
+        public void OnNext(BulkheadConfiguration value)
+        {
+            var newValue = _expression(value);
+            var hasChanged = !Equals(_currentValue, newValue);
```

Comment:
Does this work, particularly for reference types? If `T` is a `String`, what does this evaluate to if they're the same value?

Given that this is `Object.Equals` and `string`s are reference types, I'm concerned this'll do an object reference comparison.

Reply:
This should work as string has "Equals" overriden. From Object code:
```
 public virtual bool Equals(object obj)
    {
      return RuntimeHelpers.Equals(this, obj);
    }

    public static bool Equals(object objA, object objB)
    {
      if (objA == objB)
        return true;
      if (objA == null || objB == null)
        return false;
      return objA.Equals(objB);
    }
```

I've also checked with my debugger and it seems to work on my machine.

Reply:
Ah cool, thanks for checking.